FreeBSD: Add zfs jail property

ihoro opened the pull request:
Sponsored-by: SkunkWerks, GmbH

<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

It targets https://github.com/openzfs/zfs/issues/15710 feature request to add a read-only property to report name of the jail that mounted the dataset.

### Description
<!--- Describe your changes in detail -->

The decisions made:
- The `0` name is used for datasets that are not mounted or not jailed. The default zfs style is to use `-` for a missing value, but it's allowed to have a jail named like that. At the same time there is no way to name a jail as `0`: `jail: name cannot be numeric (unless it is the jid)`.
- The mechanism remembers the name only, not jid. It can be treated as a caching mechanism to avoid time complexity increase. The reason is that all properties are prepared before listing only requested ones, and traversing all prisons (with `allprison_lock`) to lookup jail name by jid does not feel good for the cases with many jailed datasets and/or jails.
- Because of the above, jail renaming is not supported -- it will still report the old name.
- Nested jails are not supported for simplicity -- only non-jailed root can see actual value of the jail property, others are provided with `0`. Such way it covers the requirement not to reveal jail name to the jail itself.
- As long as the name is cached, it's okay to have a situation when the dataset is still mounted even though the jail reported by this property is no longer present. It's not expected to be a usual case.
- The `zfsprops.7` man page was extended to outline the property. It is mentioned within the section of tunables even though it is a read-only one, it seems to be better to keep it closer to the `jailed` property.

Hence, the mechanism is simple:
- `zfs_domount()` caches `pr_name`
- `zfs_umount()` clears the name

The proposed implementation covers usual cases without extra complexity and property retrieval performance impact. The opportunities for improvements are left for the future.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

Originally the patch was created months ago for 15-CURRENT, it was re-tested for the latest 16-CURRENT. The UI/UX demonstration:

```
# cat /etc/jail.conf
jail1 {
        host.hostname = "jail1.local";
        path = "/root/tmp/jailroot_15current_1";
        persist = true;
        mount.devfs = true;
        allow.mount = true;
        allow.mount.zfs = true;
        enforce_statfs = 1;
        exec.created = "/sbin/zfs jail jail1 p3/dataset2";
        exec.start = "/bin/sh /etc/rc.d/zfs start";
        exec.stop = "/bin/sh /etc/rc.d/zfs stop";
        exec.poststop = "/sbin/zfs unjail jail1 p3/dataset2";
        children.max = 99;
}

# zpool import -d tmp/p3 p3

# zfs list -o name,mountpoint,mounted,jailed,jail
NAME         MOUNTPOINT     MOUNTED  JAILED  JAIL
p3           /p3            yes      off     0
p3/dataset1  /p3/dataset1   yes      off     0
p3/dataset2  /dataset2_mnt  no       on      0
p3/dataset3  /dataset3_mnt  no       on      0

# service jail onestart jail1

# zfs list -o name,mountpoint,mounted,jailed,jail
NAME         MOUNTPOINT     MOUNTED  JAILED  JAIL
p3           /p3            yes      off     0
p3/dataset1  /p3/dataset1   yes      off     0
p3/dataset2  /dataset2_mnt  yes      on      jail1
p3/dataset3  /dataset3_mnt  no       on      0

# jexec jail1 zfs list -o name,mountpoint,mounted,jailed,jail
NAME         MOUNTPOINT     MOUNTED  JAILED  JAIL
p3           /p3            no       off     0
p3/dataset2  /dataset2_mnt  yes      on      0

# jexec jail1 jail -c persist name=child allow.mount=true allow.mount.zfs=true enforce_statfs=1
# zfs jail jail1.child p3/dataset3
# jexec jail1.child zfs mount p3/dataset3

# jexec jail1.child zfs list -o name,mountpoint,mounted,jailed,jail
NAME         MOUNTPOINT     MOUNTED  JAILED  JAIL
p3           /p3            yes      off     0
p3/dataset3  /dataset3_mnt  yes      on      0

# jexec jail1 zfs list -o name,mountpoint,mounted,jailed,jail
NAME         MOUNTPOINT     MOUNTED  JAILED  JAIL
p3           /p3            no       off     0
p3/dataset2  /dataset2_mnt  yes      on      0

# zfs list -o name,mountpoint,mounted,jailed,jail
NAME         MOUNTPOINT     MOUNTED  JAILED  JAIL
p3           /p3            yes      off     0
p3/dataset1  /p3/dataset1   yes      off     0
p3/dataset2  /dataset2_mnt  yes      on      jail1
p3/dataset3  /dataset3_mnt  yes      on      jail1.child
```

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Quality assurance (non-breaking change which makes the code more robust against bugs)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

tonyhutter commented:
1. I saw this in the JSON output (`zfs list -j ...`):
```json
        "jail": {
          "value": "0",
          "source": {
            "type": "LOCAL",
            "data": "-"
          }
        }
```
I would have expected `type` to be NONE since this is read-only.

2. Could you update `tests/zfs-tests/tests/functional/cli_root/zfs_jail/zfs_jail_001_pos.ksh`  with some simple tests to verify `zfs list -o jail ...` works?

behlendorf commented on module/os/freebsd/zfs/zfs_vfsops.c:
```diff
@@ -1310,6 +1310,20 @@ zfs_domount(vfs_t *vfsp, char *osname)
+			zfsvfs->z_os->os_dsl_dataset->ds_jailname =
+			    kmem_zalloc(strlen(pr->pr_name) + 1, KM_SLEEP);
+			strcpy(zfsvfs->z_os->os_dsl_dataset->ds_jailname,
+			    pr->pr_name);
```
You can use `kmem_strdup(pr->pr_name);` to simplify this.

ihoro replied:
Indeed. Fixed.

behlendorf commented on module/zfs/dsl_prop.c:
```diff
@@ -1230,6 +1230,18 @@ dsl_prop_get_all_ds(dsl_dataset_t *ds, nvlist_t **nvp,
+		    (ds->ds_jailname && INGLOBALZONE(curproc)) ?
+		    ds->ds_jailname : "0"));
+		VERIFY0(nvlist_add_string(propval, ZPROP_SOURCE, setpoint));
+		VERIFY0(nvlist_add_nvlist(*nvp, "jail", propval));
```
The `fnvlist_*` wrappers should be used here, or better yet add the required error handling even if they're almost certain never to fail.

ihoro replied:
Sure. It seems I followed the examples from `dsl_prop_get_all_impl()` itself. Migrated to `fnvlist_*`, thanks.

ihoro commented:
> 1. I saw this in the JSON output (`zfs list -j ...`):
> 
> ```json
>         "jail": {
>           "value": "0",
>           "source": {
>             "type": "LOCAL",
>             "data": "-"
>           }
>         }
> ```
> 
> I would have expected `type` to be NONE since this is read-only.

Fixed.

> 
> 2. Could you update `tests/zfs-tests/tests/functional/cli_root/zfs_jail/zfs_jail_001_pos.ksh`  with some simple tests to verify `zfs list -o jail ...` works?

Sure, especially that existing test provides a working baseline. I've added a separate test not to mix multiple concerns into a single one.

amotin commented on module/os/freebsd/zfs/zfs_vfsops.c:
```diff
@@ -1310,6 +1310,18 @@ zfs_domount(vfs_t *vfsp, char *osname)
+		struct prison *pr = curthread->td_ucred->cr_prison;
+		if (pr != &prison0) {
+			zfsvfs->z_os->os_dsl_dataset->ds_jailname =
+			    kmem_strdup(pr->pr_name);
```
Again, I am not closely familiar with the zone/jail code in ZFS, but I am not sure the delegation should necessarily be tied to a mounting.  I suppose some dataset may not be mounted, but be controlled by the jail.  Am I wrong?
